<a href="https://colab.research.google.com/github/tgwon/ai4ml/blob/main/week3/5-subsetting_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Subsetting the data

## About the Data
In this notebook, we will be working with earthquake data from September 18, 2018 - October 13, 2018 (obtained from the US Geological Survey (USGS) using the [USGS API](https://earthquake.usgs.gov/fdsnws/event/1/))

## Setup
We will be working with the `data/earthquakes.csv` file again, so we need to handle our imports and read it in.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
cd /content/drive/MyDrive/Hands-On-Data-Analysis-with-Pandas-2nd-edition/ch_02

/content/drive/MyDrive/Hands-On-Data-Analysis-with-Pandas-2nd-edition/ch_02


In [3]:
import pandas as pd

df = pd.read_csv('data/earthquakes.csv')

In [4]:
df.head()

alert  cdi      code                                             detail  \
0   NaN  NaN  37389218  https://earthquake.usgs.gov/fdsnws/event/1/que...   
1   NaN  NaN  37389202  https://earthquake.usgs.gov/fdsnws/event/1/que...   
2   NaN  4.4  37389194  https://earthquake.usgs.gov/fdsnws/event/1/que...   
3   NaN  NaN  37389186  https://earthquake.usgs.gov/fdsnws/event/1/que...   
4   NaN  NaN  73096941  https://earthquake.usgs.gov/fdsnws/event/1/que...   

       dmin  felt    gap           ids   mag magType  ...  sources     status  \
0  0.008693   NaN   85.0  ,ci37389218,  1.35      ml  ...     ,ci,  automatic   
1  0.020030   NaN   79.0  ,ci37389202,  1.29      ml  ...     ,ci,  automatic   
2  0.021370  28.0   21.0  ,ci37389194,  3.42      ml  ...     ,ci,  automatic   
3  0.026180   NaN   39.0  ,ci37389186,  0.44      ml  ...     ,ci,  automatic   
4  0.077990   NaN  192.0  ,nc73096941,  2.16      md  ...     ,nc,  automatic   

            time                          title  tsunami        type  \
0  1539475168010  M 1.4 - 9km NE of Aguanga, CA        0  earthquake   
1  1539475129610  M 1.3 - 9km NE of Aguanga, CA        0  earthquake   
2  1539475062610  M 3.4 - 8km NE of Aguanga, CA        0  earthquake   
3  1539474978070  M 0.4 - 9km NE of Aguanga, CA        0  earthquake   
4  1539474716050  M 2.2 - 10km NW of Avenal, CA        0  earthquake   

                                               types     tz        updated  \
0         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475395144   
1         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475253925   
2  ,dyfi,focal-mechanism,geoserve,nearby-cities,o... -480.0  1539536756176   
3         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475196167   
4  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0  1539477547926   

                                                 url  
0  https://earthquake.usgs.gov/earthquakes/eventp...  
1  https://earthquake.usgs.gov/earthquakes/eventp...  
2  https://earthquake.usgs.gov/earthquakes/eventp...  
3  https://earthquake.usgs.gov/earthquakes/eventp...  
4  https://earthquake.usgs.gov/earthquakes/eventp...  

[5 rows x 26 columns]

In [5]:
df.columns

Index(['alert', 'cdi', 'code', 'detail', 'dmin', 'felt', 'gap', 'ids', 'mag',
       'magType', 'mmi', 'net', 'nst', 'place', 'rms', 'sig', 'sources',
       'status', 'time', 'title', 'tsunami', 'type', 'types', 'tz', 'updated',
       'url'],
      dtype='object')

In [6]:
df.shape

(9332, 26)

In [ ]:
df.loc[3:10,['mag','magType','type']]

mag magType        type
3   0.44      ml  earthquake
4   2.16      md  earthquake
5   2.61      md  earthquake
6   1.70      ml  earthquake
7   1.13      md  earthquake
8   0.92      md  earthquake
9   4.70      mb  earthquake
10  0.50      ml  earthquake

In [ ]:
df.columns.get_loc('mag')

8

In [ ]:
df.columns.get_loc('magType')

9

In [ ]:
df.columns.get_loc('type')

21

In [ ]:
df.iloc[3:10,[8,9,21]]

mag magType        type
3  0.44      ml  earthquake
4  2.16      md  earthquake
5  2.61      md  earthquake
6  1.70      ml  earthquake
7  1.13      md  earthquake
8  0.92      md  earthquake
9  4.70      mb  earthquake

In [ ]:
df.loc[(df.place.str.contains('Arizona')),['mag','magType','type']]

mag magType        type
1062  1.23      ml  earthquake
3965  0.94      md  earthquake

In [ ]:
df.sort_values('mag',ascending = False)['mag'][:10]

5263    7.5
837     7.0
4363    6.7
118     6.7
799     6.5
816     6.2
838     6.1
5323    6.1
9211    6.0
862     6.0
Name: mag, dtype: float64

In [ ]:
df.drop('alert', axis = 1, inplace=True)

In [ ]:
df.head()

cdi      code                                             detail      dmin  \
0  NaN  37389218  https://earthquake.usgs.gov/fdsnws/event/1/que...  0.008693   
1  NaN  37389202  https://earthquake.usgs.gov/fdsnws/event/1/que...  0.020030   
2  4.4  37389194  https://earthquake.usgs.gov/fdsnws/event/1/que...  0.021370   
3  NaN  37389186  https://earthquake.usgs.gov/fdsnws/event/1/que...  0.026180   
4  NaN  73096941  https://earthquake.usgs.gov/fdsnws/event/1/que...  0.077990   

   felt    gap           ids   mag magType  mmi  ... sources     status  \
0   NaN   85.0  ,ci37389218,  1.35      ml  NaN  ...    ,ci,  automatic   
1   NaN   79.0  ,ci37389202,  1.29      ml  NaN  ...    ,ci,  automatic   
2  28.0   21.0  ,ci37389194,  3.42      ml  NaN  ...    ,ci,  automatic   
3   NaN   39.0  ,ci37389186,  0.44      ml  NaN  ...    ,ci,  automatic   
4   NaN  192.0  ,nc73096941,  2.16      md  NaN  ...    ,nc,  automatic   

            time                          title  tsunami        type  \
0  1539475168010  M 1.4 - 9km NE of Aguanga, CA        0  earthquake   
1  1539475129610  M 1.3 - 9km NE of Aguanga, CA        0  earthquake   
2  1539475062610  M 3.4 - 8km NE of Aguanga, CA        0  earthquake   
3  1539474978070  M 0.4 - 9km NE of Aguanga, CA        0  earthquake   
4  1539474716050  M 2.2 - 10km NW of Avenal, CA        0  earthquake   

                                               types     tz        updated  \
0         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475395144   
1         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475253925   
2  ,dyfi,focal-mechanism,geoserve,nearby-cities,o... -480.0  1539536756176   
3         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475196167   
4  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0  1539477547926   

                                                 url  
0  https://earthquake.usgs.gov/earthquakes/eventp...  
1  https://earthquake.usgs.gov/earthquakes/eventp...  
2  https://earthquake.usgs.gov/earthquakes/eventp...  
3  https://earthquake.usgs.gov/earthquakes/eventp...  
4  https://earthquake.usgs.gov/earthquakes/eventp...  

[5 rows x 25 columns]

In [ ]:
mag = list(df.mag)
magType = list(df.magType)

In [ ]:
mag.pop
magType.pop()

'ml'

In [ ]:
mag = pd.DataFrame(mag, columns=['mag'])

In [ ]:
magType = pd.DataFrame(magType, columns=['magType'])

In [ ]:
new_df = pd.concat([mag, magType])
new_df.head()

mag magType
0  1.35     NaN
1  1.29     NaN
2  3.42     NaN
3  0.44     NaN
4  2.16     NaN

## Selecting columns
Grab an entire column using attribute notation:

In [7]:
df.mag

0       1.35
1       1.29
2       3.42
3       0.44
4       2.16
        ... 
9327    0.62
9328    1.00
9329    2.40
9330    1.10
9331    0.66
Name: mag, Length: 9332, dtype: float64

Grab an entire column using dictionary syntax:

In [8]:
df['mag']

0       1.35
1       1.29
2       3.42
3       0.44
4       2.16
        ... 
9327    0.62
9328    1.00
9329    2.40
9330    1.10
9331    0.66
Name: mag, Length: 9332, dtype: float64

Selecting multiple columns:

In [9]:
df[['mag', 'title']]

mag                                     title
0     1.35             M 1.4 - 9km NE of Aguanga, CA
1     1.29             M 1.3 - 9km NE of Aguanga, CA
2     3.42             M 3.4 - 8km NE of Aguanga, CA
3     0.44             M 0.4 - 9km NE of Aguanga, CA
4     2.16             M 2.2 - 10km NW of Avenal, CA
...    ...                                       ...
9327  0.62      M 0.6 - 9km ENE of Mammoth Lakes, CA
9328  1.00               M 1.0 - 3km W of Julian, CA
9329  2.40  M 2.4 - 35km NNE of Hatillo, Puerto Rico
9330  1.10             M 1.1 - 9km NE of Aguanga, CA
9331  0.66             M 0.7 - 9km NE of Aguanga, CA

[9332 rows x 2 columns]

Selecting columns using list comprehensions and string operations:

In [10]:
df[
    ['title', 'time']
    + [col for col in df.columns if col.startswith('mag')]
]

title           time   mag magType
0                M 1.4 - 9km NE of Aguanga, CA  1539475168010  1.35      ml
1                M 1.3 - 9km NE of Aguanga, CA  1539475129610  1.29      ml
2                M 3.4 - 8km NE of Aguanga, CA  1539475062610  3.42      ml
3                M 0.4 - 9km NE of Aguanga, CA  1539474978070  0.44      ml
4                M 2.2 - 10km NW of Avenal, CA  1539474716050  2.16      md
...                                        ...            ...   ...     ...
9327      M 0.6 - 9km ENE of Mammoth Lakes, CA  1537230228060  0.62      md
9328               M 1.0 - 3km W of Julian, CA  1537230135130  1.00      ml
9329  M 2.4 - 35km NNE of Hatillo, Puerto Rico  1537229908180  2.40      md
9330             M 1.1 - 9km NE of Aguanga, CA  1537229545350  1.10      ml
9331             M 0.7 - 9km NE of Aguanga, CA  1537228864470  0.66      ml

[9332 rows x 4 columns]

Breaking down this example:
1. the list comprehension

In [11]:
[col for col in df.columns if col.startswith('mag')]

['mag', 'magType']

2. assembling the list

In [12]:
['title', 'time'] \
+ [col for col in df.columns if col.startswith('mag')]

['title', 'time', 'mag', 'magType']

3. using this list as the list of columns

In [13]:
df[
    ['title', 'time']
    + [col for col in df.columns if col.startswith('mag')]
]

title           time   mag magType
0                M 1.4 - 9km NE of Aguanga, CA  1539475168010  1.35      ml
1                M 1.3 - 9km NE of Aguanga, CA  1539475129610  1.29      ml
2                M 3.4 - 8km NE of Aguanga, CA  1539475062610  3.42      ml
3                M 0.4 - 9km NE of Aguanga, CA  1539474978070  0.44      ml
4                M 2.2 - 10km NW of Avenal, CA  1539474716050  2.16      md
...                                        ...            ...   ...     ...
9327      M 0.6 - 9km ENE of Mammoth Lakes, CA  1537230228060  0.62      md
9328               M 1.0 - 3km W of Julian, CA  1537230135130  1.00      ml
9329  M 2.4 - 35km NNE of Hatillo, Puerto Rico  1537229908180  2.40      md
9330             M 1.1 - 9km NE of Aguanga, CA  1537229545350  1.10      ml
9331             M 0.7 - 9km NE of Aguanga, CA  1537228864470  0.66      ml

[9332 rows x 4 columns]

## Slicing
### Selecting rows
Using row numbers (inclusive of first index, exclusive of last):

In [14]:
df[100:103]

alert  cdi      code                                             detail  \
100   NaN  NaN  20280310  https://earthquake.usgs.gov/fdsnws/event/1/que...   
101   NaN  NaN  73096756  https://earthquake.usgs.gov/fdsnws/event/1/que...   
102   NaN  NaN  37388730  https://earthquake.usgs.gov/fdsnws/event/1/que...   

        dmin  felt    gap           ids   mag magType  ...  sources  \
100      NaN   NaN    NaN  ,ak20280310,  1.20      ml  ...     ,ak,   
101  0.01355   NaN  185.0  ,nc73096756,  0.59      md  ...     ,nc,   
102  0.02987   NaN   39.0  ,ci37388730,  1.33      ml  ...     ,ci,   

        status           time                                 title  tsunami  \
100  automatic  1539435449480      M 1.2 - 25km NW of Ester, Alaska        0   
101  automatic  1539435391320  M 0.6 - 8km ESE of Mammoth Lakes, CA        0   
102  automatic  1539435293090        M 1.3 - 8km ENE of Aguanga, CA        0   

           type                                              types     tz  \
100  earthquake                                  ,geoserve,origin, -540.0   
101  earthquake  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0   
102  earthquake  ,focal-mechanism,geoserve,nearby-cities,origin... -480.0   

           updated                                                url  
100  1539443551010  https://earthquake.usgs.gov/earthquakes/eventp...  
101  1539439802162  https://earthquake.usgs.gov/earthquakes/eventp...  
102  1539435940470  https://earthquake.usgs.gov/earthquakes/eventp...  

[3 rows x 26 columns]

### Selecting rows and columns with chaining

In [15]:
df[['title', 'time']][100:103]

title           time
100      M 1.2 - 25km NW of Ester, Alaska  1539435449480
101  M 0.6 - 8km ESE of Mammoth Lakes, CA  1539435391320
102        M 1.3 - 8km ENE of Aguanga, CA  1539435293090

In [16]:
df[100:103][['title', 'time']]

title           time
100      M 1.2 - 25km NW of Ester, Alaska  1539435449480
101  M 0.6 - 8km ESE of Mammoth Lakes, CA  1539435391320
102        M 1.3 - 8km ENE of Aguanga, CA  1539435293090

Order doesn't matter here:

In [17]:
df[100:103][['title', 'time']].equals(
    df[['title', 'time']][100:103]
)

True

So we know how to select rows and columns, but can we update values? Well, if we try using what we have learned so far, we will see the following warning:

In [18]:
df[110:113]['title'] = df[110:113]['title'].str.lower()

<ipython-input-18-ab67faef5099>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[110:113]['title'] = df[110:113]['title'].str.lower()


Note that it worked here, but `pandas` says we were setting a value on a copy of a slice and that we should use `loc` instead (topic of the following section):

In [19]:
df[110:113]['title']

110               M 1.1 - 35km S of Ester, Alaska
111    M 1.9 - 93km WNW of Arctic Village, Alaska
112      M 0.9 - 20km WSW of Smith Valley, Nevada
Name: title, dtype: object

## Indexing

Now if we do this with `loc` as the warning suggests, everything goes smoothly. Note we have to lower the end index by one since `loc` is inclusive of endpoints:

### loc, iloc에 인수 하나만 넣으면 행 얘기 하는 거

### df.loc[110:112, 'title'] -> 인덱스가 110, 111인 행이 출력되는 게 아니라 110부터 112까지 출력됨
### loc는 label이라서 숫자 인덱싱이 아니라 문자열 인덱싱이라고 생각

In [20]:
df.loc[110:112, 'title'] = df.loc[110:112, 'title'].str.lower()
df.loc[110:112, 'title']

110               m 1.1 - 35km s of ester, alaska
111    m 1.9 - 93km wnw of arctic village, alaska
112      m 0.9 - 20km wsw of smith valley, nevada
Name: title, dtype: object

### Indexing with `loc`
Selection of the format `loc[row_indexer, column_indexer]` where `:` can be used to select all:

In [21]:
df.loc[:,'title']

0                  M 1.4 - 9km NE of Aguanga, CA
1                  M 1.3 - 9km NE of Aguanga, CA
2                  M 3.4 - 8km NE of Aguanga, CA
3                  M 0.4 - 9km NE of Aguanga, CA
4                  M 2.2 - 10km NW of Avenal, CA
                          ...                   
9327        M 0.6 - 9km ENE of Mammoth Lakes, CA
9328                 M 1.0 - 3km W of Julian, CA
9329    M 2.4 - 35km NNE of Hatillo, Puerto Rico
9330               M 1.1 - 9km NE of Aguanga, CA
9331               M 0.7 - 9km NE of Aguanga, CA
Name: title, Length: 9332, dtype: object

We can use `loc` to select specific rows and columns without chaining. If we use row numbers with `loc`, they are now **inclusive** of the end index:

In [22]:
df.loc[10:15, ['title', 'mag']]

title   mag
10                     M 0.5 - 10km NE of Aguanga, CA  0.50
11  M 2.8 - 53km SE of Punta Cana, Dominican Republic  2.77
12                      M 0.5 - 9km NE of Aguanga, CA  0.50
13         M 4.5 - 120km SSW of Banda Aceh, Indonesia  4.50
14                   M 2.1 - 14km NW of Parkfield, CA  2.13
15       M 2.0 - 156km WNW of Haines Junction, Canada  2.00

#### Indexing with `iloc`
Exclusive of the endpoint just as Python slicing:

In [24]:
df.columns

Index(['alert', 'cdi', 'code', 'detail', 'dmin', 'felt', 'gap', 'ids', 'mag',
       'magType', 'mmi', 'net', 'nst', 'place', 'rms', 'sig', 'sources',
       'status', 'time', 'title', 'tsunami', 'type', 'types', 'tz', 'updated',
       'url'],
      dtype='object')

In [23]:
# 인덱스 8 인 컬럼 : mag -> 인덱스는 0부터 시작이니까 9번째 컬럼
df.iloc[10:15, [19, 8]]

title   mag
10                     M 0.5 - 10km NE of Aguanga, CA  0.50
11  M 2.8 - 53km SE of Punta Cana, Dominican Republic  2.77
12                      M 0.5 - 9km NE of Aguanga, CA  0.50
13         M 4.5 - 120km SSW of Banda Aceh, Indonesia  4.50
14                   M 2.1 - 14km NW of Parkfield, CA  2.13

We can use slicing syntax with `iloc` for both rows and columns:

In [25]:
df.iloc[10:15, 6:10]

gap             ids   mag magType
10   57.0    ,ci37389162,  0.50      ml
11  186.0  ,pr2018286010,  2.77      md
12   76.0    ,ci37389146,  0.50      ml
13  157.0    ,us1000hbti,  4.50      mb
14   71.0    ,nc73096921,  2.13      md

In [26]:
#magType도 포함해서 출력됨
df.loc[10:14, 'gap':'magType']

gap             ids   mag magType
10   57.0    ,ci37389162,  0.50      ml
11  186.0  ,pr2018286010,  2.77      md
12   76.0    ,ci37389146,  0.50      ml
13  157.0    ,us1000hbti,  4.50      mb
14   71.0    ,nc73096921,  2.13      md

When using `loc`, we can slice on column names. This will be inclusive of the endpoint because you can't be expected to know what the next column name will be. As such, we have multiple ways to achieve the same end goal:

In [27]:
df.iloc[10:15, 6:10].equals(
    df.loc[10:14, 'gap':'magType']
)

True

### Looking up scalar values
We used `loc` and `iloc` to grab subsets of the dataframe. However, if we are just interested in the specific value at a given `[row, column]`, then we can use `iat` and `at`. We use `at` with labels:

### 데이터프레임객체.at[인덱스번호, 컬럼이름]

In [28]:
df.at[10, 'mag']

0.5

In [30]:
df.loc[10, 'mag']

0.5

...and `iat` with integer indices:

In [29]:
df.iat[10, 8]

0.5

In [31]:
df.iloc[10,8]

0.5

## Filtering
We can filter our dataframes using a **Boolean mask**, which can be made as follows:

In [36]:
df.mag > 2

0       False
1       False
2        True
3       False
4        True
        ...  
9327    False
9328    False
9329     True
9330    False
9331    False
Name: mag, Length: 9332, dtype: bool

To use a mask for selection, we simply place it inside the brackets:

In [37]:
df[df.mag >= 7.0]

alert  cdi      code                                             detail  \
837   green  4.1  1000haa3  https://earthquake.usgs.gov/fdsnws/event/1/que...   
5263    red  8.4  1000h3p4  https://earthquake.usgs.gov/fdsnws/event/1/que...   

       dmin  felt   gap                                 ids  mag magType  ...  \
837   1.763   3.0  14.0  ,us1000haa3,pt18283003,at00pgehsk,  7.0     mww  ...   
5263  1.589  18.0  27.0             ,us1000h3p4,us1000h4p4,  7.5     mww  ...   

         sources    status           time  \
837   ,us,pt,at,  reviewed  1539204500290   
5263     ,us,us,  reviewed  1538128963480   

                                           title  tsunami        type  \
837   M 7.0 - 117km E of Kimbe, Papua New Guinea        1  earthquake   
5263           M 7.5 - 78km N of Palu, Indonesia        1  earthquake   

                                                  types     tz        updated  \
837   ,dyfi,finite-fault,general-text,geoserve,groun...  600.0  1539378744253   
5263  ,dyfi,finite-fault,general-text,geoserve,groun...  480.0  1539123134531   

                                                    url  
837   https://earthquake.usgs.gov/earthquakes/eventp...  
5263  https://earthquake.usgs.gov/earthquakes/eventp...  

[2 rows x 26 columns]

We can use masks with `loc`:

In [38]:
df.loc[
    df.mag >= 7.0,
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert  mag magType                                       title  tsunami  \
837   green  7.0     mww  M 7.0 - 117km E of Kimbe, Papua New Guinea        1   
5263    red  7.5     mww           M 7.5 - 78km N of Palu, Indonesia        1   

            type  
837   earthquake  
5263  earthquake

Masks can be created using multiple criteria when combined with bitwise operators `&` for AND and `|` for OR. We must also surround each criterion with parentheses. We can't use `and`/`or` here because we need to evaluate row by row:

In [39]:
df.loc[
    (df.tsunami == 1) & (df.alert == 'red'),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert  mag magType                              title  tsunami  \
5263   red  7.5     mww  M 7.5 - 78km N of Palu, Indonesia        1   

            type  
5263  earthquake

An example with an OR condition, which is less restrictive:

In [41]:
df.loc[
    (df.tsunami == 1) | (df.alert == 'red'),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert  mag magType                                              title  \
36      NaN  5.0     mww  M 5.0 - 165km NNW of Flying Fish Cove, Christm...   
118   green  6.7     mww            M 6.7 - 262km NW of Ozernovskiy, Russia   
501   green  5.6     mww        M 5.6 - 128km SE of Kimbe, Papua New Guinea   
799   green  6.5     mww         M 6.5 - 148km S of Severo-Kuril'sk, Russia   
816   green  6.2     mww        M 6.2 - 94km SW of Kokopo, Papua New Guinea   
...     ...  ...     ...                                                ...   
8561    NaN  5.4      mb         M 5.4 - 228km S of Taron, Papua New Guinea   
8624    NaN  5.1      mb       M 5.1 - 278km SE of Pondaguitan, Philippines   
9133  green  5.1      ml               M 5.1 - 64km SSW of Kaktovik, Alaska   
9175    NaN  5.2      mb                M 5.2 - 126km N of Dili, East Timor   
9304    NaN  5.1      mb   M 5.1 - 34km NW of Finschhafen, Papua New Guinea   

      tsunami        type  
36          1  earthquake  
118         1  earthquake  
501         1  earthquake  
799         1  earthquake  
816         1  earthquake  
...       ...         ...  
8561        1  earthquake  
8624        1  earthquake  
9133        1  earthquake  
9175        1  earthquake  
9304        1  earthquake  

[61 rows x 6 columns]

Masks can be created from any criteria that results in a Boolean. For example, we can select all earthquakes with the string `Alaska` in the `place` column with a non-null value for the `alert` column. To get non-nulls, we can use the `isnull()` method with the bitwise negation operator (`~`) or the `notnull()` method:

In [42]:
df.loc[
    (df.place.str.contains('Alaska')) & (df.alert.notnull()),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert  mag magType                                        title  \
1015  green  5.0      ml     M 5.0 - 61km SSW of Chignik Lake, Alaska   
1273  green  4.0      ml          M 4.0 - 71km SW of Kaktovik, Alaska   
1795  green  4.0      ml           M 4.0 - 60km WNW of Valdez, Alaska   
2752  green  4.0      ml         M 4.0 - 67km SSW of Kaktovik, Alaska   
3260  green  3.9      ml       M 3.9 - 44km N of North Nenana, Alaska   
4101  green  4.2      ml  M 4.2 - 131km NNW of Arctic Village, Alaska   
6897  green  3.8      ml         M 3.8 - 80km SSW of Kaktovik, Alaska   
8524  green  3.8      ml         M 3.8 - 69km SSW of Kaktovik, Alaska   
9133  green  5.1      ml         M 5.1 - 64km SSW of Kaktovik, Alaska   

      tsunami        type  
1015        1  earthquake  
1273        1  earthquake  
1795        1  earthquake  
2752        1  earthquake  
3260        0  earthquake  
4101        0  earthquake  
6897        0  earthquake  
8524        0  earthquake  
9133        1  earthquake

We can even use regular expressions here:

### r'CA|California$' : "CA" 또는 "California"로 끝나는 문자열을 검색

In [43]:
df.loc[
    (df.place.str.contains(r'CA|California$')) & (df.mag > 3.8),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert   mag magType                              title  tsunami  \
1465  green  3.83      mw  M 3.8 - 109km WNW of Trinidad, CA        0   
2414  green  3.83      mw   M 3.8 - 5km SW of Tres Pinos, CA        1   

            type  
1465  earthquake  
2414  earthquake

We can use the `between()` method to turn 2 individual checks (is less than or equal to some maximum value and is greater than or equal to some minimum value) into a single one. Note this is inclusive of the endpoint by default:

In [44]:
df.loc[
    df.mag.between(6.5, 7.5),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert  mag magType                                       title  tsunami  \
118   green  6.7     mww     M 6.7 - 262km NW of Ozernovskiy, Russia        1   
799   green  6.5     mww  M 6.5 - 148km S of Severo-Kuril'sk, Russia        1   
837   green  7.0     mww  M 7.0 - 117km E of Kimbe, Papua New Guinea        1   
4363  green  6.7     mww      M 6.7 - 263km NNE of Ndoi Island, Fiji        1   
5263    red  7.5     mww           M 7.5 - 78km N of Palu, Indonesia        1   

            type  
118   earthquake  
799   earthquake  
837   earthquake  
4363  earthquake  
5263  earthquake

We can use the `isin()` method to check for membership in a list of values:

In [45]:
df.loc[
    df.magType.isin(['mw', 'mwb']),
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert   mag magType                                   title  tsunami  \
995     NaN  3.35      mw             M 3.4 - 9km WNW of Cobb, CA        0   
1465  green  3.83      mw       M 3.8 - 109km WNW of Trinidad, CA        0   
2414  green  3.83      mw        M 3.8 - 5km SW of Tres Pinos, CA        1   
4988  green  4.41      mw       M 4.4 - 1km SE of Delta, B.C., MX        1   
6307  green  5.80     mwb  M 5.8 - 297km NNE of Ndoi Island, Fiji        0   
8257  green  5.70     mwb      M 5.7 - 175km SSE of Lambasa, Fiji        0   

            type  
995   earthquake  
1465  earthquake  
2414  earthquake  
4988  earthquake  
6307  earthquake  
8257  earthquake

We can grab the index of the minimum and maximum values of a given column and use those to select the entire row where they occur:

In [46]:
# idxmin() : mag 컬럼에서 가장 minimum한 값의 인덱스
# idmax() : mag 컬럼에서 가장 maximum한 값의 인덱스

[df.mag.idxmin(), df.mag.idxmax()]

[2409, 5263]

In [47]:
df.loc[
    [df.mag.idxmin(), df.mag.idxmax()],
    ['alert', 'mag', 'magType', 'title', 'tsunami', 'type']
]

alert   mag magType                              title  tsunami  \
2409   NaN -1.26      ml  M -1.3 - 41km ENE of Adak, Alaska        0   
5263   red  7.50     mww  M 7.5 - 78km N of Palu, Indonesia        1   

            type  
2409  earthquake  
5263  earthquake

Note that there is a `filter()` method, but it doesn't filter the data in the same sense as we discussed in this section. Here are a few things you can do with this method.

- grab columns of a dataframe by passing a list to `items`:

In [ ]:
df.filter(items=['mag', 'magType']).head()

mag magType
0  1.35      ml
1  1.29      ml
2  3.42      ml
3  0.44      ml
4  2.16      md

- grab all the columns that contain a string with the `like` parameter:

In [ ]:
df.filter(like='mag').head()

mag magType
0  1.35      ml
1  1.29      ml
2  3.42      ml
3  0.44      ml
4  2.16      md

- use regular expressions; here, we select any columns that start with `t`:

In [ ]:
df.filter(regex=r'^t').head()

time                          title  tsunami        type  \
0  1539475168010  M 1.4 - 9km NE of Aguanga, CA        0  earthquake   
1  1539475129610  M 1.3 - 9km NE of Aguanga, CA        0  earthquake   
2  1539475062610  M 3.4 - 8km NE of Aguanga, CA        0  earthquake   
3  1539474978070  M 0.4 - 9km NE of Aguanga, CA        0  earthquake   
4  1539474716050  M 2.2 - 10km NW of Avenal, CA        0  earthquake   

                                               types     tz  
0         ,geoserve,nearby-cities,origin,phase-data, -480.0  
1         ,geoserve,nearby-cities,origin,phase-data, -480.0  
2  ,dyfi,focal-mechanism,geoserve,nearby-cities,o... -480.0  
3         ,geoserve,nearby-cities,origin,phase-data, -480.0  
4  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0

- use `filter()` along the rows, by passing in `axis=0`. Here, we will use the `place` column as the index (we will cover `set_index()` in chapter 3):

In [ ]:
df.set_index('place').filter(like='Japan', axis=0).filter(items=['mag', 'magType', 'title']).head()

mag magType                                title
place                                                                        
160km NNW of Nago, Japan     4.6      mb     M 4.6 - 160km NNW of Nago, Japan
7km ESE of Asahi, Japan      5.2     mww      M 5.2 - 7km ESE of Asahi, Japan
14km E of Tomakomai, Japan   4.5     mwr   M 4.5 - 14km E of Tomakomai, Japan
139km WSW of Naze, Japan     4.7      mb     M 4.7 - 139km WSW of Naze, Japan
53km ESE of Kamaishi, Japan  4.6      mb  M 4.6 - 53km ESE of Kamaishi, Japan

This also works on `Series` objects and will run on the index:

In [ ]:
df.set_index('place').title.filter(like='Japan').head()

place
160km NNW of Nago, Japan          M 4.6 - 160km NNW of Nago, Japan
7km ESE of Asahi, Japan            M 5.2 - 7km ESE of Asahi, Japan
14km E of Tomakomai, Japan      M 4.5 - 14km E of Tomakomai, Japan
139km WSW of Naze, Japan          M 4.7 - 139km WSW of Naze, Japan
53km ESE of Kamaishi, Japan    M 4.6 - 53km ESE of Kamaishi, Japan
Name: title, dtype: object

<hr>
<div>
    <a href="./4-inspecting_dataframes.ipynb">
        <button style="float: left;">&#8592; Previous Notebook</button>
    </a>
    <a href="./6-adding_and_removing_data.ipynb">
        <button style="float: right;">Next Notebook &#8594;</button>
    </a>
</div>
<br>
<hr>